sw/device/lib/base/log.h

```diff
@@ -0,0 +1,130 @@
+ *                info, warning, error or fatal
+ *                This is indicated using LOG_TYPE_* set of macros which are
+ *                set to empty strings by default.
+ *
+ *  verbosity:    Verbosity of the message (applicable only to info messages):
+ *                none, low, medium, high, full and debug.
+ *                Based on the desired verbosity, the visibility of certain
+ *                messages can be turned off. This is expected to be done
+ *                externally (for example, if the messages go to a log,
+ *                the desired verbosity messages can be searched and filtered
+ *                out). This is indicated using the LOG_VERB_* set of macros
+ *                which are set to empty strings by default.
+ *
+ *  file name:    Name of the file using __FILE__
+ *
+ *  line number:  Line where the print originated using __LINE__
+ *
+ * Log macros support formatting specifiers; see print.h for details the subset
+ * of C specifier syntax supported.
+ *
+ * The precise mechanism for logging is dependent on the target device. On core
+ * devices, like Verilator, logs are printed using whatever |stdout| is set to
+ * in print.h. DV testbenches may use an alternative, more efficient mechanism.
+ * In DV mode, some format specifiers may be unsupported, such as %s.
+ */
+
+/**
+ * Log severities available.
+ *
+ * Additional log severities can be added as necessary.
+ */
+typedef enum log_severity {
+  kLogSeverityInfo,
+  kLogSeverityWarn,
+  kLogSeverityError,
+} log_severity_t;
+
+// Internal functions exposed only for access by macros. Their
+// real doxygen can be found in log.c.
+/**
+ * Implementation detail.
+ */
+void base_log_internal_core(log_severity_t severity, const char *file_name,
+                            uint32_t line, const char *format, ...);
+/**
+ * Implementation detail.
+ */
+void base_log_internal_dv(log_severity_t severity, const char *format, ...);
+
+/**
+ * Basic logging macro that all other logging macros delegate to.
+ *
+ * Prefer to use a LOG function with a specified severity, instead.
+ *
+ * @param severity a severity of type |log_severity_t|.
+ * @param format a format string, as described in print.h. This must be a string
+ * literal.
+ * @param ... format parameters matching the format string.
+ */
+// Currently, this simply prints directly to printf. In the future, when
+// we begin supporting DV testbenches, we can include an |if| statement here
+// that detects that using |device.h| and switches to the cheaper "dump args
+// for post process formatting" function.
+//
+// NOTE: the ##__VA_ARGS__ syntax below is a GCC/Clang extension, while
+// "" foo "" is a common C idiom to assert that a macro parameter is a string.
+#define LOG(severity, format, ...)                                       \
+  do {                                                                   \
+    if (/* device_is_dv() */ false) {                                    \
```

Comment:
Alas, `//`-comments do not work correctly inside of macros (since they break the trailing / at the end). Can I hoist it to a /**/ above the conditional?

Reply:
`/* */` are C-style comments, that's what I was suggesting. :smile:

Reply:
Done.